bdk_cbf_example/src/main.rs

```diff
@@ -0,0 +1,158 @@
+                .keychains()
+                .into_iter()
+                .map(|(k, _)| (*k, size))
+                .collect();
+            let (_, derivation_additions) = keychain_tracker
+                .txout_index
+                .reveal_to_target_multi(&targets);
+            let changeset = KeychainChangeSet {
+                derivation_indices: derivation_additions,
+                ..Default::default()
+            };
+            db.lock().unwrap().append_changeset(&changeset)?;
+            keychain_tracker.apply_changeset(changeset);
+            Ok(())
+        }
+        CbfCommands::Rescan { _batch_size } => {
+            todo!("Implement rescan from sync");
+            // This function will reveal batch_size scripts and rescan the nakamoto
+            // client
+            // If we notice, after rescan, that more than `batch_size - epsilon` scripts
+            // have been used, we rescan using 2*batch_size, and so on
+        }
+        CbfCommands::Scan => {
+            // indexing logic!
+            let mut keychain_tracker = keychain_tracker.lock().unwrap();
+
+            // find scripts!
+            let scripts = keychain_tracker
+                .txout_index
+                .revealed_spks_of_all_keychains()
+                .into_values()
+                .flatten()
+                .map(|(_, spk)| spk.clone());
+
+            let last_checkpoint = keychain_tracker.chain_graph().chain().latest_checkpoint();
+
+            let mut update = ChainGraph::<TxHeight>::default();
+
+            if let Some(last_cp) = last_checkpoint.clone() {
+                // Otherwise the chains don't connect :)
+                let _ = update.insert_checkpoint(last_cp.clone())?;
+            }
+
+            let mut derivation_additions = DerivationAdditions::default();
+
+            client.sync_setup(scripts, last_checkpoint)?;
+            loop {
+                match client.next_event()? {
+                    CbfEvent::BlockMatched(id, txs) => {
+                        let _ = update.insert_checkpoint(BlockId {
+                            height: id.height as u32,
+                            hash: id.hash,
+                        })?;
+
+                        for (tx, height) in txs {
+                            //keychain_tracker.lock().unwrap().txout_index.pad_all_with_unused(stop_gap);
+                            if keychain_tracker.txout_index.is_relevant(&tx) {
+                                println!("* adding tx to update: {} @ {}", tx.txid(), height);
+                                let _ = update.insert_tx(tx.clone(), height)?;
+                            }
+                            derivation_additions.append(keychain_tracker.txout_index.scan(&tx));
+                        }
+                    }
+                    CbfEvent::BlockDisconnected(id) => {
+                        // TODO: what happens if a block gets disconnected before I process its
+                        // filter?
```

Comment:
I really hope that `BlockMatched` doesn't get emitted for things that are not in the chain! Hopefully nakamoto guarantees this.